cgen error for `fn f() ? {}; f() or { dump(err) }`

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.11 a2f65c6, press to see full `v doctor` output</summary>

|V full version      |V 0.4.11 2cdacb8c00ddf4700b00cedfda0699719f5f12b7.a2f65c6
|:-------------------|:-------------------
|OS                  |linux, Ubuntu 20.04.6 LTS
|Processor           |4 cpus, 64bit, little endian, Intel(R) Core(TM) i3-3225 CPU @ 3.30GHz
|Memory              |0.97GB/15.05GB
|                    |
|V executable        |/home/delian/v/v
|V last modified time|2025-08-09 07:45:27
|                    |
|V home dir          |OK, value: /home/delian/v
|VMODULES            |OK, value: /home/delian/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/delian/v
|                    |
|env VFLAGS          |"-no-parallel"
|Git version         |git version 2.50.0
|V git status        |weekly.2025.32-20-ga2f65c69
|.git/config present |true
|                    |
|cc version          |cc (Ubuntu 10.5.0-1ubuntu1~20.04) 10.5.0
|gcc version         |gcc (Ubuntu 10.5.0-1ubuntu1~20.04) 10.5.0
|clang version       |N/A
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |N/A
|glibc version       |ldd (Ubuntu GLIBC 2.31-0ubuntu9.18) 2.31

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  f.v && /home/delian/v/f`
```v
fn f() ? {
	println('hello')
	return none
}

f() or { dump(err) }

```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/f.01K26XA0X0Q4YY3A3H199EV2H6.tmp.c:5094: error: assignment to void expression
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
```

**What did you expect to see?**

a compiled program, since println(err) instead of dump(err) works


> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -7106,7 +7106,7 @@ fn (mut g Gen) gen_or_block_stmts(cvar_name string, cast_typ string, stmts []ast
 								g.writeln(' }, (${option_name}*)&${cvar_name}, sizeof(${cast_typ}));')
 								g.indent--
 								return
-							} else {
+							} else if return_type.clear_option_and_result() != ast.void_type {
 								g.write('*(${cast_typ}*) ${cvar_name}${tmp_op}data = ')
 							}
 						}

--- a/vlib/v/tests/options/option_or_expr_dump_test.v
+++ b/vlib/v/tests/options/option_or_expr_dump_test.v
@@ -1,0 +1,8 @@
+fn f() ? {
+	println('hello')
+	return none
+}
+
+fn test_main() {
+	f() or { dump(err) }
+}

